warplib/src/warp_memory/arrays.cairo

```diff
@@ -0,0 +1,93 @@
+use warplib::warp_memory::WarpMemory;
+use warplib::warp_memory::WarpMemoryTrait;
+use warplib::warp_memory::WarpMemoryImpl;
+use integer::u256_from_felt252;
+use integer::u128_try_from_felt252;
+use integer::Into;
+use option::OptionTrait;
+
+trait WarpMemoryArraysTrait {
+    /// Reads the pointer of an array. If it does not exist, it creates a new one
+    fn get_or_create_id(ref self: WarpMemory, location: felt252, size: felt252) -> felt252;
+
+    /// Given the length and the felt size of an element, allocates the space
+    /// in memory and returns a pointer to it
+    fn new_dynamic_array(ref self: WarpMemory, len: felt252, elem_width: felt252) -> felt252;
+
+    /// Given a pointer to a dynamic array, the index you want to access from it
+    /// and the size of the element in such array, returns a pointer to the element
+    fn index_dyn(ref self: WarpMemory, array_ptr: felt252, index: felt252, elem_width: felt252) -> felt252;
+
+    /// Given a pointer to a static array, the index you want to access from it,
+    /// the size of the elements in the array and the it's length,  returns a pointer
+    /// to the element
+    fn index_static(ref self: WarpMemory, array_ptr: felt252, index: felt252, elem_width: felt252, length: felt252) -> felt252;
+
+    /// Given the pointer to a dynamic array, it returns it's length
+    fn length_dyn(ref self: WarpMemory, array_ptr: felt252) -> felt252;
+}
+
+
+impl WarpMemoryArraysImpl of WarpMemoryArraysTrait {
+    fn get_or_create_id(ref self: WarpMemory, location: felt252, size: felt252) -> felt252{
+        // Get the id at the location
+        let id = self.read(location);
+        if id != 0 {
+            return id;
+        }
+
+        // If no id was found create a new one
+        let id = self.alloc(size);
+        self.unsafe_write(location, id);
+        id
+    }
+
+    fn new_dynamic_array(ref self: WarpMemory, len: felt252, elem_width: felt252) -> felt252 {
```

Comment:
I think we can find a better name, since it's not adding elements instantly. Shall we name it `allocate_dynamic_array`? Maybe `declare_dynamic_array` or `reserve_memory_for_dynamic_array`?

Reply:
I think all of those are good names. But I don't see how are they better than the current one

Reply:
The problem is: how would you name a function that both allocates memory and writes it's contents at the same time? We should semantically differentiate between the two

Reply:
It's hard to tell what exactly `new` means then

Reply:
As i see it `new` means there is a dynamic array ready to be used

Reply:
Anyone always can analyze the code. The question is if it's easier or harder to do so. The purpose of name is to be a short description of function's functionalities.

 As I've asked how do we name a function that allocates memory and writes contents? Currently it's `wm${this.generatedFunctionsDef.size}_dynamic_array`. So e.g. `wm3_dynamic_array` would be calling `new_dynamic_array` followed by writes. Should we keep it that way?

Reply:
I agree that names are not the best sometimes. I would like to postpone naming schemes after this part of memory is complete, and we have the whole picture. Renaming here is affecting the work in #1045 since it is a follow up to this one

Reply:
Ok, I'm fine with postponing it